From b528bb99a3aea08958e7395c83b381cae2c6cccc Mon Sep 17 00:00:00 2001
From: Olivier BIDEAU <olivier.bideau@st.com>
Date: Thu, 26 Sep 2024 13:37:34 +0200
Subject: [PATCH 1291/1695] drivers: stpmic1: fix potential untreated pending
 irq

The irq handler was parsing the PMIC ITLATCH registers.
There was an unhandled case where an irq can become pending
after the ITLATCH registers have been parsed.

In this case the handler is left with a pending irq.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 108721b20a8084948f0e4ad94a38adac3d819755 drivers: stpmic1: fix potential untreated pending irq]
Signed-off-by: Olivier BIDEAU <olivier.bideau@st.com>
Change-Id: Iec9cc90212eefd6173619f650e395598bb72c840
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/408709
Domain-Review: Francois CHOLEAU <francois.choleau@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 .../plat-stm32mp1/drivers/stm32mp1_stpmic1.c  | 47 ++++++++++++-------
 1 file changed, 31 insertions(+), 16 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c
index 53afb2dc0..b3e4d9594 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_stpmic1.c
@@ -823,34 +823,49 @@ static enum itr_return stpmic1_irq_handler(struct itr_handler *handler __unused)
 {
 	uint8_t read_val = 0U;
 	unsigned int i = 0U;
+	uint8_t latch1 = 0U;
+	uint8_t latch2 = 0U;
+	uint8_t latch3 = 0U;
+	uint8_t latch4 = 0U;
 
 	FMSG("Stpmic1 irq handler");
 
 	stm32mp_get_pmic();
 
-	for (i = 0U; i < 4U; i++) {
-		if (stpmic1_register_read(ITLATCH1_REG + i, &read_val))
-			panic();
+	do {
+		for (i = 0U; i < 4U; i++) {
+			if (stpmic1_register_read(ITLATCH1_REG + i, &read_val))
+				panic();
 
-		if (read_val) {
-			struct pmic_it_handle_s *prv = NULL;
+			if (read_val) {
+				struct pmic_it_handle_s *prv = NULL;
 
-			FMSG("Stpmic1 irq pending %u: %#"PRIx8, i, read_val);
+				FMSG("Stpmic1 irq pending %u: %#"PRIx8, i,
+				     read_val);
 
-			if (stpmic1_register_write(ITCLEARLATCH1_REG + i,
-						   read_val))
-				panic();
+				if (stpmic1_register_write(ITCLEARLATCH1_REG +
+							   i, read_val))
+					panic();
 
-			SLIST_FOREACH(prv, &pmic_it_handle_list, link)
-				if ((prv->pmic_reg == ITCLEARMASK1_REG + i) &&
-				    (read_val & BIT(prv->pmic_bit))) {
-					FMSG("STPMIC1 send notif %u",
-					     prv->notif_id);
+				SLIST_FOREACH(prv, &pmic_it_handle_list, link)
+					if ((prv->pmic_reg == ITCLEARMASK1_REG +
+					     i) &&
+					    (read_val & BIT(prv->pmic_bit))) {
+						FMSG("STPMIC1 send notif %u",
+						     prv->notif_id);
 
-					notif_send_it(prv->notif_id);
+						notif_send_it(prv->notif_id);
 				}
+			}
 		}
-	}
+
+		stpmic1_register_read(ITLATCH1_REG, &latch1);
+		stpmic1_register_read(ITLATCH2_REG, &latch2);
+		stpmic1_register_read(ITLATCH3_REG, &latch3);
+		stpmic1_register_read(ITLATCH4_REG, &latch4);
+
+	} while ((latch1 != 0) || (latch2 != 0) || (latch3 != 0) ||
+		 (latch4 != 0));
 
 	stm32mp_put_pmic();
 
-- 
2.47.2

